Reflectometry prototype migrated

gpospelov opened the pull request.

jwuttke commented on GUI/refl/cmake/modules/configuration.cmake:
```diff
@@ -0,0 +1,67 @@
+# Compile options
+# -----------------------------------------------------------------------------
+
+add_compile_options($<$<CXX_COMPILER_ID:MSVC>:/MP>)
```
Much of this can be deleted for good, if DaRefl no longer is a separate CMake `project`. Remaining material should be integrated into top-level/cmake/BornAgain/*

jwuttke commented on GUI/CMakeLists.txt:
```diff
@@ -6,6 +6,9 @@ message(STATUS "Configuring BornAgain GUI")
+endif()
 add_subdirectory(ba3d)
 add_subdirectory(coregui)
 add_subdirectory(main)
```
This directory structure I'd expect if DeRefl is run as a subwindow of GUI1. My understanding was rather that DaRefl (or however we'll call it) is an independent executable that runs a distinct GUI of its own. Therefore, instead of `GUI/refl` I'd rather expect a new top-level directory `GUI2`, `DaRefl`, or whatsoever.

jwuttke replied:
moved to card

jwuttke commented on GUI/refl/source/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+add_subdirectory(darefl)
+
+add_executable(${executable_name} main.cpp)
+target_link_libraries(${executable_name} PRIVATE dareflcore)
```
Can we please take out this intermediate directoy `source` for good? In my working style, overlong paths are a considerable nuisance.

jwuttke replied:
moved to card

jwuttke commented on GUI/refl/source/darefl/dataloader/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+    parserpropertywidget.cpp
+    parserpropertywidget.h
+)
+
```
In current GUI, as well as in core, we glob sources. We should stay with that.

jwuttke replied:
moved to card

jwuttke commented on GUI/refl/source/darefl/dataloader/datahandler.cpp:
```diff
@@ -0,0 +1,46 @@
+// ************************************************************************** //
+
+#include <darefl/dataloader/datahandler.h>
+#include <darefl/dataloader/dataloader_utils.h>
```
Include paths should be relative to BornAgain top-level.

Within-project includes are in "...".

If need be, I can help out with edit scripts to make these changes.

jwuttke replied:
moved to card

jwuttke commented on GUI/refl/source/darefl/dataloader/dataloader_utils.cpp:
```diff
@@ -0,0 +1,202 @@
+#include <darefl/dataloader/dataloader_constants.h>
+#include <darefl/dataloader/dataloader_utils.h>
+#include <darefl/importdataview/graphimportdata.h>
+#include <fstream>
```
The header with same file name (except for extension) should come first. Then all other within-project include.

If within-project includes are in `"..."`, then clang-format does the sorting for us.

jwuttke replied:
rely on clang-format for this

jwuttke commented on GUI/refl/source/darefl/dataloader/dataloaderdialog.cpp:
```diff
@@ -0,0 +1,262 @@
+}
+
+//! Set list of target canvas to define entr where to import.
+
```
Please check the entire PR for uniform use of 3rd-person style in function comments.

jwuttke replied:
will do later

jwuttke commented on GUI/refl/source/darefl/dataloader/dataloaderdialog.h:
```diff
@@ -0,0 +1,72 @@
+class DataHandler;
+
+//! This is the main dialog for the data loader.
+
```
rm "This is"

jwuttke replied:
done

jwuttke commented on GUI/refl/source/darefl/dataloader/defaultparser.h:
```diff
@@ -0,0 +1,51 @@
+    accept_string_t m_isValidLineContent;
+    line_splitter_t m_line_splitter;
+    std::vector<std::string> m_rawData;
+    //!< correspondance of parsed data to original line index
```
-dence

jwuttke commented on GUI/refl/source/darefl/importdataview/dataselectorwidget.cpp:
```diff
@@ -0,0 +1,65 @@
+    , m_selectorTree(new QTreeView)
+    , m_canvasPropertyEditor(new ModelView::PropertyTreeView)
+    , m_graphPropertyEditor(new ModelView::PropertyTreeView)
+    , m_collapsibleWidget(new ModelView::CollapsibleListWidget) {
```
On superficial reading, I don't see the need for some of these private members; couldn't there scope be restricted to the constructor, just as for `layout`?

jwuttke replied:
moved to card

jwuttke commented on ThirdParty/GUI/qt-mvvm/source/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+
+add_subdirectory(libmvvm_model)
+add_subdirectory(libmvvm_viewmodel)
+add_subdirectory(libmvvm_view)
```
I'd really really appreciate if we could take out `source/libmvvm_` from the source paths. So we would just have `<BornAgain>/qt-mvvm/model` etc.

jwuttke replied:
moved to card

jwuttke commented on ing:
```diff
@@ -0,0 +1,169 @@
+./Sample/Specular/SpecularScalarStrategy.h:                                                const std::vector<complex_t>& kz) const;
+./Sample/Specular/SpecularScalarStrategy.h:                              const std::vector<complex_t>& kz) const;
+./Sample/Specular/SpecularScalarTanhStrategy.h:    virtual std::pair<complex_t, complex_t> transition(complex_t kzi, complex_t kzi1,
+./Sample/Specular/ISpecularStrategy.h:                             const std::vector<complex_t>& kz) const = 0;
```
What strange file is this?

jwuttke replied:
was added by mistake, now removed

jwuttke reviewed (changes requested):
Thank you, @gpospelov, for this impressive body of work.

As this PR is very consequential for our work in the next few years, we ought to deviate from our usual workflow here: I need to see responses to my comments before I approve the merge.

As some of my comments require work that may possibly be shared among us, one possible response to comments would be to move them to Redmine issues. We would then first merge this PR, and then work on improvements one by one.

gpospelov commented:
Thanks a lot for the review. As you have said, this is quite a chunk of work and we need time to absorb it.  It contains about 60k of LOC, 500 files, and 700 unit tests. I have merged two standalone projects into BornAgain. One of these projects (ReflPrototype) will dissolve with time in BornAgain GUI code, another one should stay at some distance from the rest to simplify later reuse. 

I see it actually as a cool sign, that BornAgain was changed in few lines of code only, and almost no changes in new projects have been required. One of the reasons for that is the well-thought structure of new projects and they already implemented `standalone` nature. Also, I should not forget to thank Randolf who has implemented a `minikernel` in DaRefl in such a way, that migration was easy.

I will start addressing some of the issues you have mentioned already in the next PRs. Other, like the removal of `DaRefl` in naming, will happen a bit later. Other adaptations, which are not urgent and touching areas of personal preferences and taste can be done even after. For the moment, I think, it is good enough to be merged as it is.

jwuttke commented on CMakeLists.txt:
```diff
@@ -52,6 +52,7 @@ option(ZERO_TOLERANCE "Terminate compilation on warnings" OFF)
 option(BORNAGAIN_COVERAGE "Build with test coverage information" OFF)
 option(BORNAGAIN_DEBUG_OPTIMIZATION "Build with debug optimization (gcc only)" OFF)
 option(BORNAGAIN_TIDY "Invokes clang-tidy" OFF)
+option(BORNAGAIN_NEWREFL "Compile reflectometry GUI prototype" OFF)
```
done

jwuttke commented on GUI/refl/cmake/scripts/testconfig.h.in:
```diff
@@ -0,0 +1,29 @@
+}
+
+#endif
+
```
moved to card

jwuttke commented on GUI/refl/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+set(CMAKE_CXX_STANDARD 17)
+
+option(DAREFL_BUMP_VERSION "Propagate version number" OFF)
+
```
moved to card

jwuttke commented on GUI/refl/cmake/scripts/darefl_version.h.in:
```diff
@@ -0,0 +1,51 @@
+
+} // namespace DaRefl
+
+#endif // DAREFL_CORE_VERSION_H
```
moved to card

jwuttke commented on GUI/refl/cmake/modules/ClangFormat.cmake:
```diff
@@ -0,0 +1,46 @@
+function(target_clangformat_setup target)
+  get_target_property(target_sources ${target} SOURCES)
+  clangformat_setup(${target_sources})
+endfunction()
```
moved to card

jwuttke commented on CMakeLists.txt:
```diff
@@ -138,6 +139,9 @@ if(BORNAGAIN_GUI)
     add_subdirectory(ThirdParty/GUI/qcustomplot)
 endif()
+if(BORNAGAIN_NEWREFL)
+    add_subdirectory(ThirdParty/GUI/qt-mvvm)
```
moved to card

rbeerwerth reviewed (approved):
nice work!
